squad/frontend/failures.py

```diff
@@ -0,0 +1,48 @@
+from django.shortcuts import render
+
+from squad.http import auth
+from squad.frontend.views import get_build
+
+
+@auth
+def failures(request, group_slug, project_slug, build_version):
+    project = request.project
+    build = get_build(project, build_version)
+    environments = project.environments.order_by("slug")
+
+    search = request.GET.get('search', '')
+
+    if search:
+        tests = build.tests.filter(
+            result=False,
+            has_known_issues=False,
+            metadata__name__icontains=search,
+        ).prefetch_related('environment')
+    else:
+        tests = build.tests.filter(
+            result=False,
+            has_known_issues=False,
+        ).prefetch_related('environment')
+
+    visited = set()
+    ufailures = [visited.add(t.full_name) or t for t in tests if t.full_name not in visited]
+    ufailures.sort(key=lambda uf: uf.full_name)
+
+    rows = {}
+    for t in tests:
+        if t.environment.slug not in rows:
+            rows[t.environment.slug] = {}
+
+        rows[t.environment.slug][t.full_name] = t
```

Comment:
as noted above, this logic here could be moved into the model as `Build.failures_with_confidence`

Reply:
Question: Is the performance of the patch OK on its own? Or do you think we really do need to change things up?

I'm not sure I follow 100%, but I created a patch that maybe follows along with this. See below.

If the patch is not what you had in mind, do you really think we need to pull all the tests (including the history) at once instead of over multiple queries?

```
... in squad/core/models.py, class Build ...

+  566     __failures__ = None
+  567
+  568     @property
+  569     def failures(self):
+  570         if self.__failures__:
+  571             return self.__failures__
+  572
+  573         failing = self.tests.filter(
+  574             result=False,
+  575             has_known_issues=False,
+  576         )
+  577
+  578         for t in failing:
+  579             limit = self.project.build_confidence_count
+  580             history = Test.objects.filter(
+  581                 metadata=t.metadata,
+  582                 environment=t.environment,
+  583                 build_id__lt=self.id,
+  584             ).order_by("-build_id")[:limit]
+  585
+  586             // t.set_confidence(history)
+  587
+  588         self.__failures = failing
+  589         return self.__failures__
```